### PR TITLE
Store read-only flag into metadata file for DiskS3

### PR DESCRIPTION
**Pull request by Jokser**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed possible not-working mutations for parts stored on S3 disk.

Detailed description / Documentation draft:
When we perform local backup of a data part (e.g. doing mutations, cloning) it does the following things with each file of the part:
1. Mark file as read-only
2. Create hard-link from that file to destination path

Marking file as read-only in DiskS3 just sets FS read-only flag for metadata file.
When we create hard-link from that file it modifies metadata file incrementing number of references. This is actually file modification operation and it breaks, because metadata file was marked as read-only right before:
https://gist.github.com/Jokser/d282052507b4592dca78b9c61d1295ca

We shouldn't mark file as read-only on FS in that case, instead we should store read-only inside metadata file and check it if we trying to re-write this file.

### CODE REVIEW

File: src/Disks/S3/DiskS3.cpp [hunk 1]
```diff
@@ -632,6 +645,12 @@ std::unique_ptr<ReadBufferFromFileBase> DiskS3::readFile(const String & path, si
 std::unique_ptr<WriteBufferFromFileBase> DiskS3::writeFile(const String & path, size_t buf_size, WriteMode mode, size_t estimated_size, size_t)
```
Comment (alexey-milovidov) [inline src/Disks/S3/DiskS3.cpp:645]: Also `truncateFile`.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: src/Disks/S3/DiskS3.cpp (hunk 1 above)
Comment (Jokser) [reply src/Disks/S3/DiskS3.cpp:645]: `truncateFile` is not implemented for DiskS3 yet.

**Comment by Jokser** [comment]
Failed test_storage_s3/test.py::test_custom_auth_headers I'm fixing there
https://github.com/ClickHouse/ClickHouse/pull/17299

**Review by alexey-milovidov — APPROVED** [review]

File: src/Disks/S3/DiskS3.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply src/Disks/S3/DiskS3.cpp:645]: Ok. It means Log, TinyLog, StripeLog cannot work on S3 in some cases.

**Pull request merged**